### PR TITLE
Add function to generate image from selected text

### PR DESCRIPTION
**Pull request by tninja**
Also did tiny tiny refactor (let me know if it make sense please)

@xenodium PTAL Thanks.

### CODE REVIEW

File: dall-e-shell.el [hunk 1]
```diff
@@ -419,6 +421,34 @@ ERROR-CALLBACK otherwise."
...
+          ((eq major-mode 'markdown-mode)
+           (insert (concat "![](" png-output ")\n" )))
+          (t (progn
+               (insert (concat png-output "\n"))
+                (insert-image
+                 (create-image png-output
+                                'png nil :width 400 :height 400)))))
+    ))
+
+(defun dall-e-shell-generate-image-from-region ()
```
Comment (xenodium) [inline dall-e-shell.el:446]: We can likely do without `dall-e-shell-generate-image` and `dall-e-shell-generate-image-from-region-0` for now (until we find more uses) and inline into a single interactive command. Something like:

```emacs-lisp
(defun dall-e-shell-insert-image-from-region-description ()
  "Generate and insert an image using current region as description."
  (interactive)
  (unless (region-active-p)
    (user-error "No active region"))
  (save-excursion
    (let* ((image-description (buffer-substring-no-properties (region-beginning) (region-end)))
           (png-output (dall-e-shell-post-prompt
                        (concat "Please generate image for the following text: " image-description))))
      (goto-char (region-end))
      (cond ((eq major-mode 'org-mode)
             (insert (concat "\n\n[[file:" png-output "]]\n")))
            ((eq major-mode 'markdown-mode)
             (insert (concat "\n\n![](" png-output ")\n" )))
            (t (progn
                 (insert (concat png-output "\n"))
                 (insert-image
                  (create-image png-output
                                'png nil :width 400 :height 400))))))))
```

File: dall-e-shell.el [hunk 2]
```diff
@@ -419,6 +421,34 @@ ERROR-CALLBACK otherwise."
...
                 "-d" (shell-maker--json-encode request-data))))
 
+(defun dall-e-shell-generate-image (prompt text)
+  "Generate image from text using DALL-E given prompt."
+  (let* ((question (concat prompt text))
+         (png-output (dall-e-shell-post-prompt question)))
+    png-output))
+
+(defun dall-e-shell-generate-image-from-region-0 (prompt)
+  "Generate image from region using DALL-E given prompt."
```
Comment (xenodium) [inline dall-e-shell.el:431]: No longer needed after consolidating functions (see last comment), but running `M-x checkdock` can help us avoid warnings in docs.

ps. Byte compiling the file is also handy to find other warnings (may not apply in your case).

File: dall-e-shell.el [hunk 3]
```diff
@@ -419,6 +421,34 @@ ERROR-CALLBACK otherwise."
...
+  (let* ((text (buffer-substring-no-properties (region-beginning) (region-end)))
```
Comment (xenodium) [inline dall-e-shell.el:432]: We should error out if there's no active region.

```emacs-lisp
  (unless (region-active-p)
    (user-error "No active region"))
```

File: dall-e-shell.el [hunk 4]
```diff
@@ -419,6 +421,34 @@ ERROR-CALLBACK otherwise."
...
+         (png-output (dall-e-shell-generate-image prompt text)))
+    ;; if the current buffer is org-mode, insert a org-mode image link; if it is markdown mode, insert a markdown image link; otherwise, just insert the image
```
Comment (xenodium) [inline dall-e-shell.el:434]: Nice comment, though the code is already fairly simple and descriptive. We can do without it.

File: dall-e-shell.el [hunk 5]
```diff
@@ -419,6 +421,34 @@ ERROR-CALLBACK otherwise."
...
+    (cond ((eq major-mode 'org-mode)
```
Comment (xenodium) [inline dall-e-shell.el:435]: Saving excursion, appending after the region, and inserting newlines before path adds a nice touch (added in proposed function below).

**Review by xenodium — CHANGES_REQUESTED** [review]
Thank you for the PR @tninja!

I've left a few minor comments.

File: dall-e-shell.el (hunk 3 above)
Comment (tninja) [reply dall-e-shell.el:432]: Thanks. Good to know.

File: dall-e-shell.el (hunk 5 above)
Comment (tninja) [reply dall-e-shell.el:435]: Thanks! Good to learn this.

File: dall-e-shell.el (hunk 1 above)
Comment (tninja) [reply dall-e-shell.el]: Make sense. Addressed.

**Review by xenodium — APPROVED** [review]
Thanks for applying the changes!

**Pull request merged**